metadata-ingestion/src/datahub/ingestion/source/cassandra/cassandra_api.py

```diff
@@ -0,0 +1,175 @@
+
+    def authenticate(self) -> Session:
+        """Establish a connection to Cassandra and return the session."""
+        try:
+            if self.config.cloud:
+                cloud_config = self.config.cloud_config
+                assert cloud_config
+                cluster_cloud_config = {
+                    "connect_timeout": cloud_config.connect_timeout,
+                    "use_default_tempdir": True,
+                    "secure_connect_bundle": cloud_config.secure_connect_bundle,
+                }
+                profile = ExecutionProfile(request_timeout=cloud_config.request_timeout)
+                auth_provider = PlainTextAuthProvider(
+                    "token",
+                    cloud_config.token,
+                )
+                cluster = Cluster(
+                    cloud=cluster_cloud_config,
+                    auth_provider=auth_provider,
+                    execution_profiles={EXEC_PROFILE_DEFAULT: profile},
+                    protocol_version=ProtocolVersion.V4,
+                )
+                session: Session = cluster.connect()
+                return session
+
+            auth_provider = None
+            ssl_context = None
+            if self.config.username and self.config.password:
+                ssl_context = SSLContext(PROTOCOL_TLSv1_2)
+                ssl_context.verify_mode = CERT_NONE
+                auth_provider = PlainTextAuthProvider(
+                    username=self.config.username, password=self.config.password
+                )
+
+            cluster = Cluster(
+                [self.config.contact_point],
+                port=self.config.port,
+                load_balancing_policy=None,
+            )
+
+            if auth_provider:
+                cluster = Cluster(
+                    [self.config.contact_point],
+                    port=self.config.port,
+                    auth_provider=auth_provider,
+                    ssl_context=ssl_context,
+                    load_balancing_policy=None,
+                )
+            session = cluster.connect()
+            return session
+        except Exception as e:
+            logger.error("Failed to authenticate to Cassandra", exc_info=True)
+            self.report.report_failure(
+                message="Failed to authenticate to Cassandra", exc=e
+            )
+            raise
+
+    def get_keyspaces(self) -> List:
+        """Fetch all keyspaces."""
+        try:
+            keyspaces = self.cassandra_session.execute(
+                CassandraQueries.GET_KEYSPACES_QUERY
+            )
+            keyspaces = sorted(
+                keyspaces,
+                key=lambda k: getattr(
+                    k, CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES["keyspace_name"]
+                ),
+            )
+            return keyspaces
+        except Exception:
+            logger.error("Failed to fetch keyspaces", exc_info=True)
+            self.report.warning(
```

Comment:
Nice ! Do we need this logging ? I believe report.warning will automatically do logging.

Reply:
this comment applies to all other methods below.

Reply:
removed

Reply:
Exception e needs to be passed in to `report_warning` in all these cases